pkg/graveler/committed/skip_prefix_iterator.go

```diff
@@ -0,0 +1,168 @@
+package committed
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/treeverse/lakefs/pkg/graveler"
+)
+
+type rangeValue struct {
+	r  *Range
+	vr *graveler.ValueRecord
+}
+
+type ImportIterator interface {
+	IsCurrentRangeBoundedByPrefix() bool
+	IsCurrentPrefixIncludedInRange() bool
+}
+
+type SkipPrefixIterator struct {
+	currentPrefixIndex int
+	prefixes           []graveler.Prefix
+	rangeIterator      Iterator
+	currentRangeValue  rangeValue
+}
+
+func (ipi *SkipPrefixIterator) Value() (*graveler.ValueRecord, *Range) {
+	return ipi.currentRangeValue.vr, ipi.currentRangeValue.r
+}
+
+func (ipi *SkipPrefixIterator) updateValue() (*graveler.ValueRecord, *Range) {
+	vr, r := ipi.rangeIterator.Value()
+	ipi.currentRangeValue = rangeValue{
+		r,
+		vr,
+	}
+	return vr, r
+}
+
+func (ipi *SkipPrefixIterator) Err() error {
+	return ipi.rangeIterator.Err()
+}
+func (ipi *SkipPrefixIterator) Close() {
+	ipi.rangeIterator.Close()
+}
+func (ipi *SkipPrefixIterator) SeekGE(id graveler.Key) {
+	ipi.rangeIterator.SeekGE(id)
+}
+
+func (ipi *SkipPrefixIterator) Next() bool {
+	hasNext := ipi.rangeIterator.Next()
+	if !hasNext {
+		return false
+	}
+	vr, r := ipi.updateValue()
+	ipi.updatePrefix()
+
+	if vr == nil && r != nil { // head of range
+		for ipi.IsCurrentRangeBoundedByPrefix() {
+			if ipi.rangeIterator.Err() != nil {
+				return false
+			}
+			hasNext = ipi.rangeIterator.NextRange()
+			if !hasNext {
+				return false
+			}
+			ipi.updateValue()
+			ipi.updatePrefix()
+		}
+	} else {
+		prefixLen := len(ipi.prefixes)
+		for vr != nil && ipi.currentPrefixIndex < prefixLen && strings.HasPrefix(vr.Key.String(), string(ipi.prefixes[ipi.currentPrefixIndex])) {
+			if ipi.rangeIterator.Err() != nil {
+				return false
+			}
+			hasNext = ipi.rangeIterator.Next()
+			if !hasNext {
+				return false
+			}
+			vr, _ = ipi.updateValue()
+			ipi.updatePrefix()
+		}
+	}
+	return true
+}
+
+func (ipi *SkipPrefixIterator) NextRange() bool {
+	hasNext := ipi.rangeIterator.NextRange()
+	if !hasNext {
+		return false
+	}
+	ipi.updateValue()
+	ipi.updatePrefix()
+
+	for ipi.IsCurrentRangeBoundedByPrefix() {
+		if ipi.rangeIterator.Err() != nil {
+			return false
+		}
+		hasNext = ipi.rangeIterator.NextRange()
+		if !hasNext {
+			return false
+		}
+		ipi.updateValue()
+		ipi.currentPrefixIndex++
```

Comment:
Why? What if the next range is also bounded by the prefix?

Reply:
True
This should be like in `Next()` and call `updatePrefix`.